File: Numbers/nextprime.c
/*
 * =====================================================================================
 *
 *       Filename:  nextprime.c
 *
 *    Description:  find primes until user stops asking
 *
 *        Version:  0.1
 *        Created:  07/05/14 22:35:25
 *       Revision:  none
 *       Compiler:  Clang
 *
 *         Author:  Matthew Pherigo (mwp), hybrid120@gmail.com
 *   Organization:  
 *
 * =====================================================================================
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <errno.h>
#include <stdbool.h>
#include <ctype.h>
#include <matt.h>

 /* prototypes */
long isprime(long num);

/*-----------------------------------------------------------------------------
 *  This program has three different modes:
 *
 *  - interactive (give numbers while user wants them)
 *      ./nextprime
 *
 *  - stream (give as many primes as user requested in command line argument)
 *      ./nextprime 42
 *
 *  - file (output requested number of primes to specified file)
 *      ./nextprime 42 primes.txt
 *-----------------------------------------------------------------------------*/

int main (int argc, char *argv[]) {
    long prime = 2, factor = 0;
    int n = 1;
    errno = 0;

    if (argc == 1) { /* Interactive mode */
        while (true) {
            factor = isprime(prime);
            while (factor != 0) {
                fprintf(stderr, "%ld is not prime! (found factor in %ld)\n", prime, factor);
                prime++;
                factor = isprime(prime);
            }
            printf("Prime #%d: %ld\n", n, prime);
            printf("Continue? (y/n) ");
            char c;
            switch(c = getChar()) {
                case 'n':
                case 'N':
                    exit(0);
                case 'y':
                case 'Y':
                    break;
                default:
                    continue;
            }
            prime++; n++;
        }
    }

    else if (argc == 2) { /* stream mode */
            int iterations;
            iterations = stringtoint(argv[1]);
            if (errno != 0) {
                fprintf(stderr, "Error in stringtoint: ");
                ERROR; exit(EXIT_FAILURE);
            }
            iterations++;
            for (n = 1; n < iterations; n++) {
                factor = isprime(prime);
                while (factor != 0) {
                    fprintf(stderr, "%ld is not prime! (found factor in %ld)\n", prime, factor);
                    prime++;
                    factor = isprime(prime);
                }
                printf("Prime #%d: %ld\n", n, prime);
                prime++;
            }
    }

    else if (argc == 3) { /* File mode */
        FILE *fp;                       /* output-file pointer */
        char *fp_file_name = argv[2];   /* output-file name    */
        int iterations;

        fp	= fopen(fp_file_name, "w"); /* open output file    */
        if (fp == NULL) {
            fprintf (stderr, "couldn't open file '%s'; %s\n",
                    fp_file_name, strerror(errno));
            exit (EXIT_FAILURE);
        }

        iterations = stringtoint(argv[1]);
        if (errno != 0) {
            fprintf(stderr, "Error in stringtoint: ");
            ERROR; fclose(fp); exit(EXIT_FAILURE);
        }
        iterations++;
        for (n = 1; n < iterations; n++) {
            factor = isprime(prime);
            while (factor != 0) {
                fprintf(fp, "# %ld / %ld\n", prime, factor);
                prime++;
                factor = isprime(prime);
            }
            fprintf(fp, "%d: %ld\n", n, prime);
            prime++;
        }
        if (fclose(fp) == EOF) {		/* close output file   */
            fprintf (stderr, "couldn't close file '%s'; %s\n",
                fp_file_name, strerror(errno));
            exit (EXIT_FAILURE);
        }
        printf("Done!\n");
    }

    else { /* 360 noscope mode */
        fprintf(stderr, "wat do\n");
        exit(360);
    }
}

long isprime(long num) {
    static int i; /* loop through possible factors of num */
    for (i = 2; (i < num); i++) {
        if (num % i == 0) {
            return i;
        }
    }
    return 0;
}

File: Numbers/nextprime.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# nextprime.py
"""
File: nextprime.py  
Author: Matthew Pherigo  
Email: hybrid120@gmail.com  
Github: github.com/mwpher  
Description: Finds prime numbers  

Usage:  

- nextprime.py | interactive
- nextprime.py 42 | output 42 primes
- nextprime.py 42 primes.txt | output 42 primes to primes.txt

"""

import sys

primes = [1, 2, 3]
numofprimes = 3
fname = "none"

if len(sys.argv) == 1:
	n = 0
elif len(sys.argv) == 2:
	n = int(sys.argv[1])
elif len(sys.argv) == 3:
	n, fname = int(sys.argv[1]), sys.argv[2]
else:
	print "BUH?!"

def nextprime():
	"""
	find the next prime and add it to primes[]
	"""
	global primes
	global numofprimes
	global fname
	prime = False
	n = primes[len(primes) - 1] + 1
	while True:
		factor = isprime(n)
		if factor == 0:
			primes.append(n)
			break
		else:
			print str(n), "is divisible by", str(factor)
			n += 1
	print str(n), "is prime"
	numofprimes += 1
def isprime(n):
	"""
	check if (n) is prime. if not, return first factor found for use by
	program.
	"""
	i = 2
	while i < n:
		if (n % i) == 0:
			return i
		else:
			i += 1
			continue
	return 0
def outputresults():
	"""
	print results to file or stdout
	"""
	global primes
	global numofprimes
	global fname
	finalstate = "I found {0}: {1}".format(numofprimes, primes)
	if fname != "none":
		with open(fname, 'w') as f:
			f.write(finalstate)
	else:
		print finalstate

if n == 0:
	"""
	if n is 0, enter interactive mode
	"""
	while True:
		nextprime()
		answer = raw_input("Continue? (Y/n)")
		if len(answer) > 0 and answer[0].lower() == 'n':
			outputresults()
			break
else:
	while numofprimes < n:
		nextprime()
	outputresults()
